### ISSUE TITLE
Windows support

### ISSUE THREAD

**Issue by ryb73** [reasonml-labs/decco#20]

**Comment by mrmurphy** [comment]
We could get windows support one of two ways:

- Windows users could use `esy` as their package manager instead of (or alongside?) NPM and install this package from source, then build the PPX themselves. That's a bummer for them though.
- Someone could contribute a change that moves the build pipeline for this repo from Travis to Github Actions and adds a Windows build and artifact.

Does this seem like a good first issue?

**Comment by ryb73** [comment]
Yeah, it'd be nice for the ppx to work the same for Windows users as it does for Mac and Linux. I've been wanting to get my Windows machine working so I can look at this but haven't gotten around to it. It'd be great to have someone work on it!

Re: github actions vs travis: I haven't looked closely at github actions yet; does it provide similar functionality to travis? If I remember correctly Travis does support Windows but I was having trouble getting it working, which is why I only included Mac and Linux. I'd be interested in exploring GitHub Actions though.

**Comment by TheSpyder** [comment]
What would be needed to contribute this? I'm starting to evaluate decco for a project but we run CI tests on Windows which means we need to be able to compile on Windows. I might be able to allocate some time to fixing this.

Would you prefer Travis for windows or github actions (which yes are similar to Travis)?

**Comment by TheSpyder** [comment]
I made some attempt to build on windows, initial failures were solved by using a newer version of OCaml. In the end I couldn't get past a `mingw` requirement. Even if I can get that working, it will really only be useful for validating the executable works. This isn't something I would know how to set up on travis or github actions.

Other projects I know that build with `esy` and offer windows builds all use azure pipelines, which are free for open source, so that's where I will focus my efforts:
https://github.com/esy-ocaml/esy-ci-scripts

**Comment by TheSpyder** [comment]
SUCCESSS!
It took a lot of wrangling, but I have a passing build with binaries for linux, mac and windows:
https://thespyder.visualstudio.com/Decco%20Builds/_build/results?buildId=25&view=artifacts&type=publishedArtifacts

The individual binaries are published by their respective builds, and then they are all brought together in the `npm-package` artifact. That could be downloaded as a zip and published as-is (in fact I'm likely to publish it to my company's internal registry while I wait for help with what to do next).

As a part of this I built a `ppx.cmd` file to make it work with the `--as-ppx` command line on windows, and I have tested it on mac and windows by manually copying the package to `node_modules/decco`.
https://github.com/TheSpyder/decco

**Comment by ryb73** [comment]
That's awesome, thanks @TheSpyder! Would you be able to do a pull request? What are the missing links we need to figure out in order to incorporate it?

**Comment by TheSpyder** [comment]
There are a few things missing:
* making sure it works on a variety of machines (I just realised I went with `macOS-latest` so the mac binary probably requires 10.15)
* an azure pipelines project owned by you (the project deletes the travis config, but the interim broken state is probably ok since the project doesn't release very often)
* figuring out how to release and publish. Looking at the Travis config, there seems to be mechanisms in place to load an API key to publish to NPM? I assume Azure has the same thing. Azure has the concept of [deployment approval gates](https://docs.microsoft.com/en-us/azure/devops/pipelines/release/deploy-using-approvals?view=azure-devops), so the job can always include a "publish this build" button for every commit that is only activated when necessary.
